File: pixsim7/backend/main/shared/content_rating.py
"""
Content Rating Utilities

Shared helpers for content rating normalization, clamping, and comparison.
Canonical scale: sfw | romantic | mature_implied | restricted

@see packages/shared/types/src/contentRating.ts - TypeScript counterpart
"""

from typing import Optional, List

# Canonical content rating order (least to most permissive)
RATING_ORDER: List[str] = ["sfw", "romantic", "mature_implied", "restricted"]


def get_rating_index(value: str) -> int:
    """
    Get the index of a content rating in the hierarchy.
    Returns 0 (sfw) if value is not recognized.

    Args:
        value: Content rating string

    Returns:
        Index in RATING_ORDER (0-3), defaults to 0 for unknown values
    """
    try:
        return RATING_ORDER.index(value)
    except ValueError:
        return 0


def normalize_rating(value: Optional[str]) -> str:
    """
    Normalize a content rating value to the canonical scale.
    Returns 'sfw' for None, empty, or unrecognized values.

    Args:
        value: Raw content rating string (may be None or invalid)

    Returns:
        Normalized content rating from RATING_ORDER
    """
    if not value or value not in RATING_ORDER:
        return "sfw"
    return value


def clamp_rating(
    value: str,
    world_max: Optional[str] = None,
    user_max: Optional[str] = None
) -> str:
    """
    Clamp content rating to the most restrictive of world and user maximums.

    Args:
        value: Proposed content rating
        world_max: World's maximum allowed rating (optional)
        user_max: User's maximum allowed rating (optional)

    Returns:
        Clamped rating (most restrictive of value, world_max, user_max)

    Example:
        >>> clamp_rating("mature_implied", "romantic", None)
        "romantic"
        >>> clamp_rating("sfw", "restricted", "romantic")
        "sfw"
    """
    # Normalize input
    rating = normalize_rating(value)
    rating_idx = get_rating_index(rating)

    # Start with maximum possible index
    effective_max_idx = len(RATING_ORDER) - 1

    # Apply world constraint
    if world_max and world_max in RATING_ORDER:
        effective_max_idx = min(effective_max_idx, get_rating_index(world_max))

    # Apply user constraint
    if user_max and user_max in RATING_ORDER:
        effective_max_idx = min(effective_max_idx, get_rating_index(user_max))

    # Clamp rating to max
    if rating_idx > effective_max_idx:
        return RATING_ORDER[effective_max_idx]

    return rating


def is_rating_allowed(value: str, max_rating: str) -> bool:
    """
    Check if a content rating is allowed within a maximum constraint.

    Args:
        value: Content rating to check
        max_rating: Maximum allowed rating

    Returns:
        True if value is at or below max_rating in the hierarchy

    Example:
        >>> is_rating_allowed("romantic", "mature_implied")
        True
        >>> is_rating_allowed("restricted", "romantic")
        False
    """
    value_idx = get_rating_index(normalize_rating(value))
    max_idx = get_rating_index(normalize_rating(max_rating))
    return value_idx <= max_idx

File: packages/shared/logic-core/src/contentRating.ts
/**
 * Content Rating Utilities
 *
 * Shared helpers for content rating normalization, clamping, and comparison.
 * Canonical scale: sfw | romantic | mature_implied | restricted
 *
 * @see pixsim7/backend/main/shared/content_rating.py - Python counterpart
 */

import type { ContentRating } from '@pixsim7/shared.types';

/**
 * Canonical content rating order (least to most permissive)
 */
export const CONTENT_RATING_ORDER = [
  'sfw',
  'romantic',
  'mature_implied',
  'restricted',
] as const;

/**
 * Get the index of a content rating in the hierarchy.
 * Returns 0 (sfw) if value is not recognized.
 *
 * @param value - Content rating string
 * @returns Index in CONTENT_RATING_ORDER (0-3), defaults to 0 for unknown values
 */
export function getContentRatingIndex(value: string | undefined): number {
  if (!value) return 0;
  const index = CONTENT_RATING_ORDER.indexOf(value as ContentRating);
  return index >= 0 ? index : 0;
}

/**
 * Normalize a content rating value to the canonical scale.
 * Returns 'sfw' for undefined, empty, or unrecognized values.
 *
 * @param value - Raw content rating string (may be undefined or invalid)
 * @returns Normalized content rating
 */
export function normalizeContentRating(value?: string): ContentRating {
  if (!value) return 'sfw';
  const index = CONTENT_RATING_ORDER.indexOf(value as ContentRating);
  return index >= 0 ? (value as ContentRating) : 'sfw';
}

/**
 * Clamp content rating to the most restrictive of world and user maximums.
 *
 * @param value - Proposed content rating
 * @param worldMax - World's maximum allowed rating (optional)
 * @param userMax - User's maximum allowed rating (optional)
 * @returns Clamped rating (most restrictive of value, worldMax, userMax)
 *
 * @example
 * clampContentRating("mature_implied", "romantic", undefined)
 * // => "romantic"
 *
 * clampContentRating("sfw", "restricted", "romantic")
 * // => "sfw"
 */
export function clampContentRating(
  value: string,
  worldMax?: string,
  userMax?: string
): ContentRating {
  const rating = normalizeContentRating(value);
  const ratingIndex = getContentRatingIndex(rating);

  // Start with maximum possible index
  let effectiveMaxIndex = CONTENT_RATING_ORDER.length - 1;

  // Apply world constraint
  if (worldMax) {
    const worldIndex = getContentRatingIndex(worldMax);
    effectiveMaxIndex = Math.min(effectiveMaxIndex, worldIndex);
  }

  // Apply user constraint
  if (userMax) {
    const userIndex = getContentRatingIndex(userMax);
    effectiveMaxIndex = Math.min(effectiveMaxIndex, userIndex);
  }

  // Clamp rating to max
  if (ratingIndex > effectiveMaxIndex) {
    return CONTENT_RATING_ORDER[effectiveMaxIndex];
  }

  return rating;
}

/**
 * Check if a content rating is allowed within a maximum constraint.
 *
 * @param value - Content rating to check
 * @param maxRating - Maximum allowed rating
 * @returns True if value is at or below maxRating in the hierarchy
 *
 * @example
 * isContentRatingAllowed("romantic", "mature_implied")
 * // => true
 *
 * isContentRatingAllowed("restricted", "romantic")
 * // => false
 */
export function isContentRatingAllowed(value: string, maxRating: string): boolean {
  const valueIndex = getContentRatingIndex(normalizeContentRating(value));
  const maxIndex = getContentRatingIndex(normalizeContentRating(maxRating));
  return valueIndex <= maxIndex;
}
